### DISCUSSION TITLE
Problems writing last week's semantic relatedness tutorial data to SQL?

### DISCUSSION THREAD

**Discussion by legatrix** [jspsych/jsPsych#892] [Q&A] [answered]
Hi,

While my n-back task from the YouTube tutorial successfully writes every time, the semantic relatedness audio task from last week refuses to write any data to my SQL database. As far as I can see, the file structure on the server is exactly analogous. No error messages anywhere. Little bit stuck on how to move forward. Thanks in advance for any suggestions.

In an unrelated problem, the task will not run on Safari (it freezes at the first fixation cross of the practice), but it will on Firefox.

**Comment by jodeleeuw** [comment, marked as answer]
Do you have any error messages from SQL? I think you found a helpful one last time.

**Reply by legatrix** [reply]
Hi Josh,

I'm pretty sure that error message was in the console, and this time there are none in the console. I'll try poking around some more.

**Reply by legatrix** [reply]
The only warning I get is "Cookie “PHPSESSID” will be soon treated as cross-site cookie against “https://bradley.fit/l/semanticRelatedness/php/save_data.php” because the scheme does not match", but there were similar warnings for every other call of an external file (all the scripts, all the audio files).

**Reply by legatrix** [reply]
There is now an error log in the php folder with the save script (not sure why it wasn't created last time but hey). The message is:

"PHP Warning:  count(): Parameter must be an array or an object that implements Countable in [...]/save_data.php on line 39"

Not sure what to do with that. Once again, the col names have all been checked to exactly match the output of `jsPsych.data.get().uniqueNames()`.

**Reply by legatrix** [reply]
This is a long shot, but do you think it could be anything to do with the output datapoint:

`"view_history": "[{\"page_index\":0,\"viewing_time\":1272},{\"page_index\":1,\"viewing_time\":873},{\"page_index\":2,\"viewing_time\":2072}]"`

**Reply by jodeleeuw** [reply]
I've saved `view_history` data to MySQL using this script before, so I don't think that's the issue. Did you double check that you are passing the correct data to `save_data.php`? I've made a mistake where I send JSON instead of an array or vice versa.

**Reply by legatrix** [reply]
Thanks for the quick reply! The end of my script has exactly the same as the previous tutorial, calling the `serverComm` script:

```
		var save = {
            type: 'call-function',
            func: function(){
                jsPsych.data.displayData();
                serverComm.save_data(jsPsych.data.get().values());
            }
        }
        timeline.push(save);

        jsPsych.init({
            timeline: timeline,
            preload_audio: all_audio,
            show_preload_progress_bar: true,
            use_webaudio: true, // ensure true if running from server
        })

    </script>
</html>
```

so I don't see what could be going wrong there.

**Reply by legatrix** [reply]
I have just noticed one thing when inspecting the .values() in the console after the experiment: there are a few `null`s knocking around, for example for RTs on fixation trials. I seem to remember that, with the very simple n-back example, if there were any `null` RTs, nothing saved to SQL.

**Reply by jodeleeuw** [reply]
In your MySQL table do you allow for NULL responses? that's an option for columns.

**Reply by legatrix** [reply]
Yes! The man, the legend. What great advice. Clearly I am an SQL noob but still, glad to have it sorted!
